### PR TITLE
feat: customize staking module for greenfield

### PR DESCRIPTION
**Pull request by yutianwu**
### Description

This pr contains the changes from the following prs:

+ https://github.com/bnb-chain/greenfield-cosmos-sdk/pull/46
+ https://github.com/bnb-chain/greenfield-cosmos-sdk/pull/47
+ https://github.com/bnb-chain/greenfield-cosmos-sdk/pull/79
+ https://github.com/bnb-chain/greenfield-cosmos-sdk/pull/100
+ https://github.com/bnb-chain/greenfield-cosmos-sdk/pull/107

### Rationale

Customize staking module for greenfield

### Example

N/A

### Changes

Notable changes:
* customize staking module for greenfield

### CODE REVIEW

**Comment by forcodedancing** [comment]
This pr is also included in staking module, can we also include it?
https://github.com/bnb-chain/greenfield-cosmos-sdk/pull/129/files

File: x/authz/keeper/keeper.go [hunk 1]
```diff
@@ -103,7 +108,7 @@ func (k Keeper) DispatchActions(ctx sdk.Context, grantee sdk.AccAddress, msgs []
 
 			grant, found := k.getGrant(ctx, skey)
 			if !found {
-				return nil, errorsmod.Wrapf(authz.ErrNoAuthorizationFound, "failed to update grant with key %s", string(skey))
+				return nil, errorsmod.Wrapf(authz.ErrNoAuthorizationFound, "failed to Update grant with key %s", string(skey))
```
Comment (guagualvcha) [inline x/authz/keeper/keeper.go:111]: still use low case `update`

File: x/slashing/keeper/msg_server.go [hunk 2]
```diff
@@ -53,3 +55,50 @@ func (k msgServer) Unjail(goCtx context.Context, msg *types.MsgUnjail) (*types.M
...
+	// Jail the validator if not already jailed. This will begin unbonding the
+	// validator if not already unbonding (tombstoned).
+	if !validator.IsJailed() {
+		k.Jail(ctx, consAddr)
+	}
+
+	// Jail forever.
+	k.JailForever(ctx, consAddr)
+
+	// todo: confirm the event deleted here
```
Comment (guagualvcha) [inline x/slashing/keeper/msg_server.go:93]: Who need confirm this? @KeefeL ?

File: x/staking/keeper/msg_server.go [hunk 3]
```diff
@@ -40,15 +42,36 @@ func (k msgServer) CreateValidator(goCtx context.Context, msg *types.MsgCreateVa
 		return nil, err
 	}
 
+	delAddr, err := sdk.AccAddressFromHexUnsafe(msg.DelegatorAddress)
```
Comment (guagualvcha) [inline x/staking/keeper/msg_server.go:45]: Can you check the `DelegatorAddress` field, as the latets version said it is deprecated, and never use it, and also `The validator address bytes and delegator address bytes refer to the same account while creating validator`.

File: x/staking/keeper/msg_server.go [hunk 4]
```diff
@@ -210,6 +310,8 @@ func (k msgServer) Delegate(goCtx context.Context, msg *types.MsgDelegate) (*typ
 		return nil, err
 	}
 
+	// TODO: And a hard fork to allow all delegations, before that fork, only self delegation allowed.
```
Comment (guagualvcha) [inline x/staking/keeper/msg_server.go:313]: @j75689  @KeefeL  please note.

**Review by guagualvcha — APPROVED** [review]

**Pull request merged**